### PR TITLE
ddl: add alter index visibility DDL support

### LINKED ISSUE #9246: UCP: Support  Invisible Indexes
## Description

MySQL supports [invisible indexes](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html); that is, indexes that are not used by the optimizer. 

This is a useful feature when you want to drop an index in a safe way. Invisible indexes make it possible to test the effect of removing an index on query performance, without making a destructive change that must be undone should the index turn out to be required. Dropping and re-adding an index can be expensive for a large table, whereas making it invisible and visible are fast, in-place operations.

Support the option of `VISIBLE | INVISIBLE`.
```
CREATE [...] INDEX index_name
    [index_type]
    ON tbl_name (key_part,...)
    [index_option]

index_option:
    {VISIBLE | INVISIBLE}
```

Also, consider the following:
* Display information about invisible indexes in the output of INFORMATION_SCHEMA.STATISTICS or SHOW INDEX.
* Index hints need to report errors when used in invisible indexes.
* The primary key cannot be set to an invisible index.

## Score 
* 1294

## Mentor(s)
* @zimulala

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills
* DDL
* Golang

## Learning Materials
MySQL's [document](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) for syntax reference.
[TiDB DDL architecture](https://github.com/pingcap/tidb/blob/master/docs/design/2018-10-08-online-DDL.md)

### Time

GanttStart: 2020-07-10
GanttDue: 2020-08-15
GanttProgress: 100%

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve?

Issue Number: close #9246 

Problem Summary: Support Invisible Indexes

### What is changed and how it works?

Proposal: #15338 

What's Changed:

This PR add new DDL statement support:
```sql
alter table t alter index idx invisible;
alter table t alter index idx visible;
```

How it Works:

Add a new DDL job `AlterTableIndexInvisible`.

### Related changes

- None

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Support invisible indexes

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -310,6 +310,32 @@ func onRenameIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return ver, nil
 }
 
+func validateAlterIndexVisibility(indexName model.CIStr, invisible bool, tbl *model.TableInfo) (bool, error){
+	if idx := tbl.FindIndexByName(indexName.L); idx == nil {
+		return false, errors.Trace(infoschema.ErrKeyNotExists.GenWithStackByArgs(indexName.O, tbl.Name))
+	} else {
+		if idx.Invisible == invisible {
```
Comment (zimulala) [inline ddl/index.go:317]:
```suggestion
		else if idx.Invisible == invisible {
```

File: ddl/db_integration_test.go [hunk 2]
```diff
@@ -2257,3 +2257,30 @@ func (s *testIntegrationSuite3) TestCreateTableWithAutoIdCache(c *C) {
...
+	tk.MustQuery(query).Check(testkit.Rows("a NO", "b YES"))
+
+	tk.MustExec("alter table t alter index b invisible")
+	tk.MustQuery(query).Check(testkit.Rows("a NO", "b NO"))
+
+	tk.MustGetErrMsg("alter table t alter index non_exists_idx visible", "[schema:1176]Key 'non_exists_idx' doesn't exist in table 't'")
+
+	// Alter implicit primary key to invisible index should throw error
+	tk.MustExec("create table t1(a int NOT NULL, unique(a))")
+	// TODO: error here
```
Comment (zimulala) [inline ddl/db_integration_test.go:2284]: Could we remove this comment or complete this comment？

Comment (Deardrops) [reply ddl/db_integration_test.go:2284]: After #15421 merge, I will uncomment there.

**Comment by Deardrops** [comment]
> Could we need to update the result of `show create table`?

This job has finished in #15366

File: ddl/rollingback.go [hunk 3]
```diff
@@ -349,7 +349,8 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
 		model.ActionModifyColumn, model.ActionAddForeignKey,
 		model.ActionDropForeignKey, model.ActionRenameTable,
 		model.ActionModifyTableCharsetAndCollate, model.ActionTruncateTablePartition,
-		model.ActionModifySchemaCharsetAndCollate, model.ActionRepairTable, model.ActionModifyTableAutoIdCache:
+		model.ActionModifySchemaCharsetAndCollate, model.ActionRepairTable,
+		model.ActionModifyTableAutoIdCache, model.ActionAlterIndexVisibility:
```
Comment (zimulala) [inline ddl/rollingback.go:353]: Please add a test to `TestCancelJob` in ddl_worker_test.go.

File: ddl/index.go [hunk 4]
```diff
@@ -310,6 +310,30 @@ func onRenameIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return ver, nil
 }
 
+func validateAlterIndexVisibility(indexName model.CIStr, invisible bool, tbl *model.TableInfo) (bool, error){
+	if idx := tbl.FindIndexByName(indexName.L); idx == nil {
+		return false, errors.Trace(infoschema.ErrKeyNotExists.GenWithStackByArgs(indexName.O, tbl.Name))
+	} else if idx.Invisible == invisible {
+			return true, nil
+		}
```
Comment (zimulala) [inline ddl/index.go:318]: This line may make gofmt unhappy.

File: ddl/db_integration_test.go [hunk 5]
```diff
@@ -2257,3 +2257,30 @@ func (s *testIntegrationSuite3) TestCreateTableWithAutoIdCache(c *C) {
...
+	tk.MustExec("alter table t alter index a invisible")
+	tk.MustQuery(query).Check(testkit.Rows("a NO", "b NO"))
+
+	tk.MustExec("alter table t alter index b visible")
+	tk.MustQuery(query).Check(testkit.Rows("a NO", "b YES"))
+
+	tk.MustExec("alter table t alter index b invisible")
+	tk.MustQuery(query).Check(testkit.Rows("a NO", "b NO"))
+
+	tk.MustGetErrMsg("alter table t alter index non_exists_idx visible", "[schema:1176]Key 'non_exists_idx' doesn't exist in table 't'")
```
Comment (djshow832) [inline ddl/db_integration_test.go:2280]: What about altering expression indexes?

**Comment by Deardrops** [comment]
@djshow832 @zimulala All comment addressed, PTAL

File: ddl/db_integration_test.go [hunk 6]
```diff
@@ -2257,3 +2257,43 @@ func (s *testIntegrationSuite3) TestCreateTableWithAutoIdCache(c *C) {
...
+
+	// Alter expression index
+	tk.MustExec("create table t3(a int NOT NULL, b int)")
+	tk.MustExec("alter table t3 add index idx((a+b));")
+	query = queryIndexOnTable("t3")
+	tk.MustQuery(query).Check(testkit.Rows("idx YES"))
+
+	tk.MustExec("alter table t3 alter index idx invisible")
+	tk.MustQuery(query).Check(testkit.Rows("idx NO"))
+}
```
Comment (djshow832) [inline ddl/db_integration_test.go:2299]: What about altering primary key? (where pkIsHandle=true)

Comment (Deardrops) [reply ddl/db_integration_test.go:2303]: Done, PTAL @djshow832

File: ddl/ddl_worker_test.go [hunk 7]
```diff
@@ -524,7 +527,17 @@ func checkColumnsNotFound(t table.Table, colNames []string) bool {
 	return notFound
 }
 
-func (s *testDDLSuite) TestCancelJob(c *C) {
+func checkIdxVisibility(changedTable table.Table, idxName string, expected bool) (flag bool) {
+	for _, idxInfo := range changedTable.Meta().Indices {
+		if idxInfo.Name.O == idxName && idxInfo.Invisible == expected {
+			flag = true
+			break
```
Comment (zimulala) [inline ddl/ddl_worker_test.go:534]: Could we use `return true` directly? Then we can remove line533-534.

File: ddl/ddl_worker_test.go [hunk 8]
```diff
@@ -524,7 +527,17 @@ func checkColumnsNotFound(t table.Table, colNames []string) bool {
...
+		}
+	}
+	return
+}
+
+func (s *testDDLSuite) TestCancelJob1(c *C) {
```
Comment (zimulala) [inline ddl/ddl_worker_test.go:540]: Why change this name?

Comment (Deardrops) [reply ddl/ddl_worker_test.go]: My Mistake, fixed, PTAL @zimulala

File: planner/core/prepare_test.go [hunk 9]
```diff
@@ -748,3 +748,36 @@ func (s *testPrepareSuite) TestPrepareForGroupByMultiItems(c *C) {
...
+	tk.MustExec(`prepare stmt1 from "select a from t order by a"`)
+
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
+	c.Assert(len(tk.Se.GetSessionVars().StmtCtx.IndexNames), Equals, 1)
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.IndexNames[0], Equals, "t:idx_a")
+
+	tk.MustExec("alter table t alter index idx_a invisible")
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1"))
```
Comment (zimulala) [inline planner/core/prepare_test.go:775]: Why this statement needs to be executed twice?

Comment (Deardrops) [reply planner/core/prepare_test.go:775]: https://github.com/pingcap/tidb/blob/014d101a5d7fbd78f4bc306251181c01c0f01c7c/planner/core/prepare_test.go#L74-L76
we need to execute sql at least twice to test the prepare statement cache.

Comment (zimulala) [reply planner/core/prepare_test.go:775]: Got it!

Comment (zimulala) [reply planner/core/prepare_test.go:775]: Got it!

File: ddl/ddl_api.go [hunk 10]
```diff
@@ -4771,3 +4773,36 @@ func (d *ddl) DropSequence(ctx sessionctx.Context, ti ast.Ident, ifExists bool)
...
+		invisible = true
+	}
+
+	skip, err := validateAlterIndexVisibility(indexName, invisible, tb.Meta())
+	if skip {
+		return nil
+	}
+	if err != nil {
+		return errors.Trace(err)
+	}
```
Comment (AilinKid) [inline ddl/ddl_api.go:4794]: here should check error first

File: bindinfo/bind_test.go [hunk 11]
```diff
@@ -1147,6 +1147,8 @@ func (s *testSuite) TestInvisibleIndex(c *C) {
 	tk.MustGetErrMsg(
 		"create global binding for select * from t using select * from t use index(idx_b) ",
 		"[planner:1176]Key 'idx_b' doesn't exist in table 't'")
+
+	// create bind using index
```
Comment (AilinKid) [inline bindinfo/bind_test.go:1151]:
```suggestion
	// Create bind using index
```

File: bindinfo/bind_test.go [hunk 12]
```diff
@@ -1158,14 +1160,13 @@ func (s *testSuite) TestInvisibleIndex(c *C) {
...
 
-	// TODO: Add these test
-	//tk.MustExec("alter table t alter index idx_a invisible")
-	//tk.MustQuery("select * from t where a > 3")
-	//c.Assert(tk.Se.GetSessionVars().StmtCtx.IndexNames[0], Equals, "t:idx_a")
-	//c.Assert(tk.MustUseIndex("select * from t where a > 3", "idx_a(a)"), IsTrue)
-	//
-	//tk.MustExec("execute stmt1")
-	//c.Assert(len(tk.Se.GetSessionVars().StmtCtx.IndexNames), Equals, 0)
+	// and then make this index invisible
```
Comment (AilinKid) [inline bindinfo/bind_test.go:1163]:
```suggestion
	// And then make this index invisible
```

File: ddl/index.go [hunk 13]
```diff
@@ -712,6 +736,31 @@ func checkRenameIndex(t *meta.Meta, job *model.Job) (*model.TableInfo, model.CIS
...
+	if err := job.DecodeArgs(&indexName, &invisible); err != nil {
+		job.State = model.JobStateCancelled
+		return nil, indexName, invisible, errors.Trace(err)
+	}
+
+	_, err = validateAlterIndexVisibility(indexName, invisible, tblInfo)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, indexName, invisible, errors.Trace(err)
+	}
```
Comment (AilinKid) [inline ddl/index.go:760]: the `skip` value should be kept, and do the double-check here.
if index passes the first skip-check in ddl_api, put into the queue for running. there may be a job before it in the queue has already changed it to invisible

**Pull request merged**